File: app_manager/appmanager/src/main/java/app/editors/manager/managers/retrofit/ConverterFactory.java
package app.editors.manager.managers.retrofit;

import com.google.gson.GsonBuilder;

import java.lang.annotation.Annotation;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.reflect.Type;

import lib.toolkit.base.managers.utils.TimeUtils;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.converter.simplexml.SimpleXmlConverterFactory;

public class ConverterFactory extends Converter.Factory {

    @Retention(RetentionPolicy.RUNTIME)
    public @interface Json {
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface Xml {
    }

    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
        for (Annotation annotation : annotations) {
            if (annotation.annotationType() == Xml.class) {
                return SimpleXmlConverterFactory.create().responseBodyConverter(type, annotations, retrofit);
            } else if (annotation.annotationType() == Json.class) {
                return getGsonConverter().responseBodyConverter(type, annotations, retrofit);
            }
        }
        return getGsonConverter().responseBodyConverter(type, annotations, retrofit);
    }

    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
        for (Annotation annotation : methodAnnotations) {
            if (annotation.annotationType() == Xml.class) {
                return SimpleXmlConverterFactory.create().requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit);
            } else if (annotation.annotationType() == Json.class) {
                return getGsonConverter().requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit);
            }
        }
        return getGsonConverter().requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit);
    }

    private Converter.Factory getGsonConverter() {
        return GsonConverterFactory.create(new GsonBuilder()
                .setLenient()
                .setPrettyPrinting()
                .setDateFormat(TimeUtils.OUTPUT_PATTERN_DEFAULT)
                .serializeNulls()
                .create());
    }
}

File: app_manager/core/src/main/java/app/documents/core/network/webdav/ConverterFactory.kt
@file:Suppress("DEPRECATION")

package app.documents.core.network.webdav

import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.converter.simplexml.SimpleXmlConverterFactory
import java.lang.reflect.Type

class ConverterFactory : Converter.Factory() {
    @Retention(AnnotationRetention.RUNTIME)
    annotation class Json

    @Retention(AnnotationRetention.RUNTIME)
    annotation class Xml

    override fun responseBodyConverter(
        type: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<ResponseBody, *>? {
        for (annotation in annotations) {
            if (annotation.annotationClass == Xml::class) {
                return SimpleXmlConverterFactory.create().responseBodyConverter(type, annotations, retrofit)
            } else if (annotation.annotationClass == Json::class) {
                return jsonConverter.responseBodyConverter(type, annotations, retrofit)
            }
        }
        return jsonConverter.responseBodyConverter(type, annotations, retrofit)
    }

    override fun requestBodyConverter(
        type: Type,
        parameterAnnotations: Array<Annotation>,
        methodAnnotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<*, RequestBody>? {
        for (annotation in methodAnnotations) {
            if (annotation.annotationClass == Xml::class) {
                return SimpleXmlConverterFactory.create()
                    .requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit)
            } else if (annotation.annotationClass == Json::class) {
                return jsonConverter.requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit)
            }
        }
        return jsonConverter.requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit)
    }

    @Suppress("JSON_FORMAT_REDUNDANT")
    private val jsonConverter: Converter.Factory
        get() = Json {
            isLenient = true
            ignoreUnknownKeys = true
        }.asConverterFactory("application/json".toMediaType())
}